dev/DynamicDependency/PackageDependency.cpp

```diff
@@ -0,0 +1,199 @@
+    packageDependency.m_packageFamilyName = object.GetNamedString(winrt::hstring(L"PackageFamilyName"));
+    packageDependency.m_minVersion = PACKAGE_VERSION{ uint64_from_string(object.GetNamedString(winrt::hstring(L"MinVersion")).c_str()) };
+    packageDependency.m_packageDependencyProcessorArchitectures = static_cast<MddPackageDependencyProcessorArchitectures>(uint32_from_string(object.GetNamedString(winrt::hstring(L"ProcessorArchitectures")).c_str()));
+    packageDependency.LifetimeKind(object.GetNamedString(winrt::hstring(L"LifetimeKind")));
+    packageDependency.m_lifetimeArtifact = object.GetNamedString(winrt::hstring(L"LifetimeArtifact"));
+    packageDependency.m_options = static_cast<MddCreatePackageDependencyOptions>(uint32_from_string(object.GetNamedString(winrt::hstring(L"Options")).c_str()));
+
+    return packageDependency;
+}
+
+MddCore::PackageDependency MddCore::PackageDependency::FromJSON(PCSTR jsonUtf8)
+{
+    return FromJSON(utf8_to_hstring(jsonUtf8));
+}
+
+bool MddCore::PackageDependency::IsExpired() const
+{
+    switch (m_lifetimeKind)
+    {
+    case MddPackageDependencyLifetimeKind::Process:
+        return false;
+    case MddPackageDependencyLifetimeKind::FilePath:
+        return !std::filesystem::exists(m_lifetimeArtifact);
+    case MddPackageDependencyLifetimeKind::RegistryKey:
+        return !IsRegistryKeyExists(m_lifetimeArtifact);
+    default:
+        FAIL_FAST_HR(E_UNEXPECTED);
+    }
+}
+
+bool MddCore::PackageDependency::IsRegistryKeyExists(
+    const std::wstring& key)
+{
+    PCWSTR subkey{};
+    auto root{ ParseRegistryKey(key, subkey) };
+
+    wil::unique_hkey hkey;
+    auto rc{ ::RegOpenKeyExW(root, subkey, 0, KEY_READ, wil::out_param(hkey)) };
+    if (rc == ERROR_SUCCESS)
+    {
+        return true;
+    }
+    else if ((rc == ERROR_FILE_NOT_FOUND) || (rc == ERROR_PATH_NOT_FOUND))
+    {
+        return false;
+    }
+    THROW_HR_MSG(HRESULT_FROM_WIN32(rc), "RegOpenKey failed %d", rc);
+}
+
+HKEY MddCore::PackageDependency::ParseRegistryKey(
+    const std::wstring& key,
+    PCWSTR& subkey)
+{
+    size_t offset{};
+    auto root{ ParseRegistryKey(key, offset) };
+    subkey = key.c_str() + offset;
+    return root;
+}
+
+HKEY MddCore::PackageDependency::ParseRegistryKey(
+    const std::wstring& key,
+    size_t& offsetToSubkey)
+{
+    HKEY root{};
+    auto offset{ key.find(L'\\') };
+    THROW_HR_IF_MSG(E_INVALIDARG, offset == std::wstring::npos, "Invalid RegistryKey %ls", key.c_str());
+    THROW_HR_IF_MSG(E_INVALIDARG, offset == 0, "Invalid RegistryKey %ls", key.c_str());
+    auto prefix{ key.substr(0, offset) };
+    if (prefix == L"HKCR")
+    {
+        root = HKEY_CLASSES_ROOT;
+    }
+    else if (prefix == L"HKCU")
+    {
+        root = HKEY_CURRENT_USER;
+    }
+    else if (prefix == L"HKLM")
+    {
+        root = HKEY_LOCAL_MACHINE;
+    }
+    else if (prefix == L"HKU")
+    {
+        root = HKEY_USERS;
+    }
+    else
```

Comment:
[External]
Is it okay that this function doesn't handle fully specified registry hive names, e.g. `HKEY_LOCAL_MACHINE`? Haven't fully traced this back to see if it's accessible via public APIs.

Reply:
Per spec [6.1. Win32 API - MsixDynamicDependency.hpp](https://github.com/microsoft/ProjectReunion/blob/main/specs/dynamicdependencies/DynamicDependencies.md#61-win32-api---msixdynamicdependencyhpp)
```
enum class MddPackageDependencyLifetimeKind
{
...
    /// The lifetime artifact is a registry key in the format
    /// 'root\\subkey' where root is one of the following: HKLM, HKCU, HKCR, HKU.
    /// The package dependency is implicitly deleted when this is deleted.
    RegistryKey = 2,
```

We can add more root names if we like e.g. HKEY_LOCAL_MACHINE handled equivalent to HKLM. Are the long names desirable? If so please open a new Issue for follow up